### PR TITLE
DOC: Improved documentation for jax.numpy.kron and jax.numpy.outer

### PR DESCRIPTION
**Pull request by selamw1**
Part of https://github.com/google/jax/issues/21461

### CODE REVIEW

File: jax/_src/numpy/lax_numpy.py [hunk 1]
```diff
@@ -7405,9 +7405,32 @@ def inner(
...
 
-@util.implements(np.outer, skip_params=['out'])
 @partial(jit, inline=True)
 def outer(a: ArrayLike, b: ArrayLike, out: None = None) -> Array:
+  """Compute the outer product of two arrays.
+
+  JAX implementation of :func:`numpy.outer`.
+
+  Args:
+    a: first input array.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7415]: Mention that `a` and `b` will be flattened if they are multi-dimensional

File: jax/_src/numpy/lax_numpy.py [hunk 2]
```diff
@@ -7405,9 +7405,32 @@ def inner(
...
+    b: second input array.
+    out: unsupported by JAX.
+
+  Returns:
+    The outer product of the inputs `a` and `b`.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7420]: Mention that the shape of the output is `(a.size, b.size)`.

File: jax/_src/numpy/lax_numpy.py [hunk 3]
```diff
@@ -7443,9 +7466,33 @@ def cross(a, b, axisa: int = -1, axisb: int = -1, axisc: int = -1,
   return moveaxis(c, 0, axisc)
 
 
-@util.implements(np.kron)
 @jit
 def kron(a: ArrayLike, b: ArrayLike) -> Array:
+  """Compute the Kronecker product of two input arrays.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7471]: I'm not sure how familiar the Kronecker product is. We may want to describe the computation it does

File: jax/_src/numpy/lax_numpy.py [hunk 4]
```diff
@@ -7443,9 +7466,33 @@ def cross(a, b, axisa: int = -1, axisb: int = -1, axisc: int = -1,
...
+
+  JAX implementation of :func:`numpy.kron`.
+
+  Args:
+    a: first input array.
+    b: second input array.
+
+  Returns:
+    A new array representing the Kronecker product of the inputs `a` and `b`.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7480]: Mention the shape of the output is the elementwise product of the input shapes.

File: jax/_src/numpy/lax_numpy.py [hunk 5]
```diff
@@ -7443,9 +7466,33 @@ def cross(a, b, axisa: int = -1, axisb: int = -1, axisc: int = -1,
...
-@util.implements(np.kron)
 @jit
 def kron(a: ArrayLike, b: ArrayLike) -> Array:
+  """Compute the Kronecker product of two input arrays.
+
+  JAX implementation of :func:`numpy.kron`.
+
+  Args:
+    a: first input array.
+    b: second input array.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7477]: Mention that the arrays may be of any shape

File: jax/_src/numpy/lax_numpy.py (hunk 1 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 2 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 3 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py:7472]: Kronecker product description is added.

File: jax/_src/numpy/lax_numpy.py (hunk 4 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: Done

File: jax/_src/numpy/lax_numpy.py (hunk 5 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: Done

File: jax/_src/numpy/lax_numpy.py [hunk 6]
```diff
@@ -7405,9 +7405,33 @@ def inner(
...
+  JAX implementation of :func:`numpy.outer`.
+
+  Args:
+    a: first input array, if not 1D it will be flattened.
+    b: second input array, if not 1D it will be flattened.
+    out: unsupported by JAX.
+
+  Returns:
+    The outer product of the inputs ``a`` and ``b``. Returned array
+    will be of shape (``a.shape``, ``a.shape``).
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7421]: I don't think this is correct; I believe the returned array will be of shape `(a.size, b.size)` (please confirm)

File: jax/_src/numpy/lax_numpy.py [hunk 7]
```diff
@@ -7443,9 +7467,40 @@ def cross(a, b, axisa: int = -1, axisb: int = -1, axisc: int = -1,
...
 def kron(a: ArrayLike, b: ArrayLike) -> Array:
+  """Compute the Kronecker product of two input arrays.
+
+  JAX implementation of :func:`numpy.kron`.
+
+  The Kronecker product is an operation on two matrices of arbitrary size that
+  produces a block matrix. Each element of the first matrix ``a`` is multiplied by
+  the entire second matrix ``b``. If ``a`` has shape (m, n) and ``b`` has shape (p, q),
+  the resulting matrix will have shape (m * p, n * q). The blocks of the resulting
+  matrix are formed by multiplying each element of ``a`` by the entire matrix ``b``.
```
Comment (jakevdp) [inline jax/_src/numpy/lax_numpy.py:7480]: This statement is repeated twice; I'd remove this second one.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/numpy/lax_numpy.py (hunk 6 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: corrected.

File: jax/_src/numpy/lax_numpy.py (hunk 7 above)
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: done.

File: jax/_src/numpy/lax_numpy.py [hunk 8]
```diff
@@ -7405,9 +7405,33 @@ def inner(
...
+  JAX implementation of :func:`numpy.outer`.
+
+  Args:
+    a: first input array, if not 1D it will be flattened.
+    b: second input array, if not 1D it will be flattened.
+    out: unsupported by JAX.
+
+  Returns:
+    The outer product of the inputs ``a`` and ``b``. Returned array
+    will be of shape (``a.size``, ``b.size``).
```
[the remark this answers is not in the document]
Comment (selamw1) [reply jax/_src/numpy/lax_numpy.py]: format fixed.

**Review by jakevdp — APPROVED** [review]
Looks great, thanks!

**Pull request merged**